File: forge/src/main/java/org/valkyrienskies/clockwork/forge/content/curiosities/tools/gravitron/tool/GravitronToolBase.java
package org.valkyrienskies.clockwork.forge.content.curiosities.tools.gravitron.tool;

import com.mojang.blaze3d.vertex.PoseStack;
import com.simibubi.create.foundation.render.SuperRenderTypeBuffer;
import com.simibubi.create.foundation.utility.RaycastHelper;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.core.BlockPos;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraft.world.phys.Vec3;
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronItem;
import org.valkyrienskies.clockwork.forge.ClockworkModForge;
import org.valkyrienskies.clockwork.forge.content.curiosities.tools.gravitron.GravitronHandler;
import org.valkyrienskies.clockwork.mixinduck.MixinPlayerDuck;

public abstract class GravitronToolBase implements IGravitronTool {

    protected GravitronHandler gravitronHandler = null;
    public BlockPos clickedPos;
    public Vec3 clickedLocation;
    public static byte GRAB = 1;
    public static byte ASSEMBLE = 2;
    public static byte GRABSSEMBLE = 3;

    public void updateTargetPos() {
        LocalPlayer player = Minecraft.getInstance().player;

        BlockHitResult trace = RaycastHelper.rayTraceRange(player.level(), player, player.getBlockReach() + 2);
        if (trace == null || trace.getType() != HitResult.Type.BLOCK) {
            return;
        }

        clickedPos = trace.getBlockPos().immutable();
        clickedLocation = clickedPos.getCenter();
    }

    @Override
    public boolean handleRightClick() {
        return false;
    }

    @Override
    public void init() {
        gravitronHandler = ClockworkModForge.GRAVITRON_HANDLER;
    }

    @Override
    public void renderTool(PoseStack ms, SuperRenderTypeBuffer buffer, Vec3 camera) {

    }

    @Override
    public void renderOverlay(GuiGraphics graphics, float partialTicks, int width, int height) {

    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/curiosities/tools/gravitron/tool/GravitronToolBase.kt
package org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.tool

import com.mojang.blaze3d.vertex.PoseStack
import com.simibubi.create.content.schematics.client.tools.ToolType
import com.simibubi.create.foundation.render.SuperRenderTypeBuffer
import com.simibubi.create.foundation.utility.RaycastHelper
import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec3
import org.valkyrienskies.clockwork.ClockworkPackets
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronHandler
import org.valkyrienskies.clockwork.platform.SharedValues
import org.valkyrienskies.mod.common.util.toDoubles
import org.valkyrienskies.mod.common.util.toJOMLF

abstract class GravitronToolBase : IGravitronTool {
    protected var gravitronHandler: GravitronHandler? = null
    var clickedPos: BlockPos? = null
    var clickedLocation: Vec3? = null


    fun updateTargetPos() {
        val player = Minecraft.getInstance().player

        val trace = RaycastHelper.rayTraceRange(
            player!!.level, player, 15.0
        )
        if (trace == null || trace.type != HitResult.Type.BLOCK) {
            return
        }

        clickedPos = trace.blockPos.immutable()

        clickedLocation = clickedPos!!.toDoubles().add(0.5,0.5,0.5)
    }

    override fun handleRightClick(): Boolean {
        return false
    }

    override fun handleMouseWheel(delta: Double): Boolean {
        return false;
    }

    override fun init() {
        gravitronHandler = SharedValues.gravitronHandler
    }

    override fun renderTool(ms: PoseStack?, buffer: SuperRenderTypeBuffer?, camera: Vec3?) {
    }

    override fun renderOverlay(poseStack: PoseStack, partialTicks: Float, width: Int, height: Int) {
    }

    companion object {
        @JvmField
        var GRAB: Byte = 1

        @JvmField
        var ASSEMBLE: Byte = 2

        @JvmField
        var GRABSSEMBLE: Byte = 3
    }
}